### PR TITLE
Emit global uTP payload event in uTP listener when an uTP stream is closed/reset

### PR DESCRIPTION
**Pull request by ogenev**
### What was wrong?

`UtpListener` runs as a self-contained service in the background and handles all active uTP streams.
We need a way to process all closed uTP streams and pass the payload to the overlay service, so we can verify, store and propagate gossip/POKE the received content via uTP.

Subtask of https://github.com/ethereum/trin/issues/313.

### How was it fixed?

1. Rename `UtpSocket` to `UtpStream`.
2. Refactor the way we are storing the received payload (DATA packets) in the uTP socket.
3.  Add a new `AddActiveConnection` UtpListener request and move the initialization of an uTP stream inside UtpListener.
4.  Add `UtpStream` -> `UtpListener` event channel and emit event inside `UtpSocket` when stream state changes to `Closed` or `Reset`.
5. Upon every `UtpStream` event, `UtpListener` processes this event (doing some simple active streams maintenance by removing closed or reset streams) and emit a global `UtpListenerEvent`
6. Remove redundant and dead code.
7. Add a test case for uTP listener events.

Manually tested OFFER/ACCEPT scenarios with Fluffy.

### To-Do
- [x] 1. Add protocol id when creating uTP socket and move uTP socket creation outside of UtpListener.
- [x] 2. Add UtpSocket <-> UtpListener event channel and emit an event upon uTP socket closing.
- [x] 3. Emit global UtpListener event with uTP payload, stream id, and protocol id on closed uTP socket. 

The next two tasks will be addressed in the next PR (https://github.com/ethereum/trin/pull/337) as this is becoming quite large:

4. Add a uTP listener event handler to dispatch the global event to the specific overlay networks.
5. Listen and handle dispatched events in the overlay layer.

[//]: # (Stay ahead of things, add list items here!)
- [x] Add entry to the [release notes](https://github.com/ethereum/trin/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

File: trin-core/src/portalnet/types/content_key.rs
Comment (ogenev) [reply trin-core/src/portalnet/types/content_key.rs:10]: SSZ encoded Bytelist, maybe it is better to use `Bytelist` instead of this type alias?

Comment (njgheorghita) [reply trin-core/src/portalnet/types/content_key.rs:10]: Hmmm, either way works for me. Maybe just update the description for the type alias

**Review by njgheorghita — APPROVED** [review]
👍

**Review by jacobkaufmann — COMMENTED** [review]
I'm generally opposed to this sort of approach, where we send messages to `UtpListener` to "ask" for closed uTP streams. In my view, it adds both unnecessary complexity and unnecessary resource consumption.

The approach I would suggest is similar to the way `Discv5` emits `Discv5Event` to a channel that we can stream. Upon socket close, `UtpListener` can emit some message to a channel where the message contains the socket info and any data that the socket received. `OverlayService` can listen to this channel to handle those messages.

**Comment by ogenev** [comment]
> The approach I would suggest is similar to the way `Discv5` emits `Discv5Event` to a channel that we can stream. Upon socket close, `UtpListener` can emit some message to a channel where the message contains the socket info and any data that the socket received. `OverlayService` can listen to this channel to handle those messages.

Yeah, this seems a more robust way to handle the closed streams.
However, this will require an event channel per network because if we emit a global overlay event, then only one of the overlay networks will consume it because we will run multiple overlay protocols. 

Also, the connection state is closed inside `UtpSocket` and I think at least for now, we will still need to process all active sockets in a set time interval in `UtpListener` to check which one is closed and then emit the event. Another alternative is to add event channels between `UtpSocket` and `UtpListener` but this will bring more complexity.

One major issue I realized today is that now `UtpListener` doesn't recognize which stream for which network is related and we can't emit specific events for a specific network.

I will try to refactor a bit and first store all active uTP connections in `UtpListener` in a separate `HashMap` per network and then try to pass state and history event receivers to the `OverlayService` struct and emit an event when a connection is closed (after `UtpListener` process all active connections to check which one is closed).

 As we are using the same `OverlayService` struct for every network, probably I should pass those event receivers as `Option<UnboundedReceiver<...>>`, because we may not always want to run all the networks, do you have any better idea?
 
 Let me know what you think and if this makes sense to you!

**Comment by ogenev** [comment]
Good, I updated the PR description with some TODOs and will re-open it for review when all of them are done.

File: trin-core/src/utp/stream.rs
Comment (mrferris) [inline trin-core/src/utp/stream.rs:122]:
```suggestion
    /// Request to initiate and add uTP stream to the active connections
```

File: trin-core/src/utp/trin_helpers.rs
Comment (mrferris) [inline trin-core/src/utp/trin_helpers.rs:54]:
```suggestion
/// Used to track which stream an overlay request corresponds with
```

File: trin-core/src/utp/trin_helpers.rs
Comment (mrferris) [inline trin-core/src/utp/trin_helpers.rs:56]: The name `UtpStreamId` is a bit confusing to me, I think because `ConnId` is already acting as an identifier for streams.
I wonder if `UtpStreamType` or `UtpStreamOrigin` would be a more descriptive name.

File: trin-core/src/utp/stream.rs
Comment (mrferris) [inline trin-core/src/utp/stream.rs:123]: I wonder if `InitiateConnection` would be a more descriptive name for this event.

The name `AddActiveConnection` is a little bit confusing to me because it seems like both `Connect` and `AddActiveConnection` events add an active connection to the `utp_connections`. In fact, `Connect` seems like it creates more of an "active" connection than `AddActiveConnection` does, because by the time a `Connect` event happens we necessarily have two participants in the stream, whereas an `AddActiveConnection` might only ever result in an inactive connection if the other peer never connects.

**Review by mrferris — COMMENTED** [review]
👍 to the architecture decision & implementation that resulted from the above conversation.

I left some suggestions about comments and naming. Let me know if I've misunderstood anything.

File: trin-core/src/utp/stream.rs
Comment (jacobkaufmann) [inline trin-core/src/utp/stream.rs:284]: Why the minus one here?

File: trin-core/src/utp/stream.rs
Comment (jacobkaufmann) [inline trin-core/src/utp/stream.rs:310]: Same question as above regarding the minus one.

File: trin-core/src/utp/stream.rs
Comment (jacobkaufmann) [inline trin-core/src/utp/stream.rs:71]: I sort of prefer no alias, so that it's clear that we are passing around a `Arc<RwLock<..>>`, and because we only use this once as far as I can tell. Just a personal preference though.

**Review by jacobkaufmann — COMMENTED** [review]
Looking good. I feel that there are some comments that we ought to address before merge, but most of my comments are minor or questions about things I don't quite understand. Would like to get clarity on those questions before approval.

In the future, we should minimize refactor changes like naming that add a lot of non-logical line changes to the diff. If possible, deferring those changes can make the PR less noisy.

File: trin-core/src/utp/stream.rs
Comment (ogenev) [reply trin-core/src/utp/stream.rs:114]: Changed this to `Request to create and connect to a uTP stream initiated by a remote node` as this is used to create an uTP stream AND to connect to the remote peer.

File: trin-core/src/utp/trin_helpers.rs
Comment (ogenev) [reply trin-core/src/utp/trin_helpers.rs:56]: `ConnId` is not really an identifier for streams in the portal network context, it is identifying all the packets that belong to the same connection. Each uTP socket has one connection ID for sending packets and a different connection ID for receiving packets.

Anyways, I'm happy to rename this if it will improve code readability, what is your opinion @jacobkaufmann?

File: trin-core/src/utp/stream.rs
Comment (ogenev) [reply trin-core/src/utp/stream.rs:123]: You are right, the idea of `AddActiveConnection` is to create an uTP stream and start listening for incoming SYN packets.
`Connect` is creating an uTP stream and is sending SYN packet for a handshake with the remote peer.

File: trin-core/src/portalnet/overlay_service.rs
Comment (ogenev) [reply trin-core/src/portalnet/overlay_service.rs:431]: Nice catch! We should check the overlay routing table here.  Added an overlay `find_enr` method similar to discv5's.

I think it is ok to do the check inside this method as we need the `Enr` only for sending uTP packets over discv5 as part of the uTP stream. We are calling it from `handle_find_content` and `handle_offer` which have knowledge only for the `NodeId` of the request.

File: trin-core/src/portalnet/overlay_service.rs
Comment (ogenev) [reply trin-core/src/portalnet/overlay_service.rs:50]: Renamed. It is configurable with a constant for now :smile_cat:. Seriously, not sure if we need to add this to the CLI but I think it should not be part of this PR.

File: trin-core/src/portalnet/overlay_service.rs
Comment (ogenev) [reply trin-core/src/portalnet/overlay_service.rs:558]: It is just a nasty hack to make the code compile without refactoring the current hacky `RecursiveFindContent` json-rpc endpoint as it is going to be reimplemented soon. It is not expected currently to work recursively with uTP anyways, so we are not breaking anything that is working now :)

File: trin-core/src/portalnet/overlay_service.rs
Comment (ogenev) [reply trin-core/src/portalnet/overlay_service.rs:445]: FIxed.

File: trin-core/src/portalnet/overlay.rs
Comment (ogenev) [reply trin-core/src/portalnet/overlay.rs]: It seems `rustfmt` is formating it this way...

File: trin-core/src/utp/stream.rs
Comment (ogenev) [reply trin-core/src/utp/stream.rs:284]: This is because we are expecting to receive a SYN packet with the random connection id (conn_id) generated by the acceptor node (us) and this is the value we are using to create `ConnectionKey` to add the stream to the active connections hashmap (when handling OFFER message for example).

However, when we receive DATA or FIN packets, we are expecting a packet to have `conn_id + 1` value (according to uTP [specs](https://www.bittorrent.org/beps/bep_0029.html)). So if we want to check if the packet is part of an active uTP stream, when handling those packets, we have to generate a `ConnectionKey` using the` packet.connection_id - 1`.

The good point here is that we should handle the overflow and use `wrapped_sub(1)`

File: trin-core/src/utp/stream.rs
Comment (jacobkaufmann) [reply trin-core/src/utp/stream.rs:69]: Hmm, that's annoying. maybe we can use

```rust
pub type UnboundedUtpEventReceiver = UnboundedReceiver<UtpListenerEvent>;
```

and then use `Arc<RwLock<UnboundedUtpEventReceiver>>`.

If not, then this is okay.

File: trin-core/src/portalnet/overlay_service.rs
Comment (jacobkaufmann) [reply trin-core/src/portalnet/overlay_service.rs]: We would probably also add the "ENR cache" to the search after we implement that.

File: trin-core/src/utp/stream.rs
Comment (jacobkaufmann) [reply trin-core/src/utp/stream.rs]: Hmm I see. Is there a test case that we can add for this logic?

File: trin-core/src/portalnet/overlay.rs
Comment (ogenev) [reply trin-core/src/portalnet/overlay.rs]: Ah, I didn't know that this is what you want, fixed.

File: trin-core/src/utp/stream.rs
Comment (ogenev) [reply trin-core/src/utp/stream.rs:69]: Actually, we don't need to wrap this in Arc<RwLock<..>> anymore, so I removed the alias :partying_face:.

File: trin-core/src/portalnet/overlay_service.rs
Comment (ogenev) [reply trin-core/src/portalnet/overlay_service.rs]: Not sure if we have a use case to make it configurable right now but this line is in the diff because when changing the code in the previous version of the PR, I decided to extract it for better readability. I'll be more careful when renaming or making such unrelated changes in the future :innocent: .

This constant is about the overlay bucket refresh lookup logic which is out of the scope of this PR.

File: trin-core/src/utp/stream.rs
Comment (ogenev) [reply trin-core/src/utp/stream.rs]: Added an integration test for the uTP listener events that catches such types of changes.

File: trin-core/src/utp/trin_helpers.rs
Comment (mrferris) [reply trin-core/src/utp/trin_helpers.rs:56]: Yea sorry I wasn't making sense by conflating the terms stream and connection.

Passing something non-numeric called `Id` in the same calls as a numeric `Id` is what seemed off to me.

Though looking back, I think I'm splitting hairs. I'm good with leaving it as it is.

File: trin-core/src/utp/stream.rs
Comment (ogenev) [reply trin-core/src/utp/stream.rs:230]: No need for this at the moment, spawning a new task here requires additional refactoring and adds more complexity.  If it causes performance issues, we will change this in the future.

**Pull request merged**
